Source/DafnyCore/Resolver/PreType/Flows.cs

```diff
@@ -0,0 +1,443 @@
+    var value = PreTypeConstraints.Pad(AdjustableType.ToStringAsBottom(sink), 20);
+    output.WriteLine($"    {bound}  {value}    {TokDescription()}");
+  }
+}
+
+abstract class FlowIntoExpr : Flow {
+  private readonly Type sink;
+
+  protected FlowIntoExpr(Type sink, IToken tok, string description = "")
+    : base(tok, description) {
+    this.sink = AdjustableType.NormalizeSansAdjustableType(sink);
+  }
+
+  protected FlowIntoExpr(Expression sink, IToken tok, string description = "")
+    : base(sink.tok, description) {
+    this.sink = sink.UnnormalizedType;
+  }
+
+  protected abstract Type GetSourceType();
+
+  public override bool Update(FlowContext context) {
+    return UpdateAdjustableType(sink, GetSourceType(), context);
+  }
+
+  public override void DebugPrint(TextWriter output) {
+    if (sink is AdjustableType adjustableType) {
+      var sourceType = GetSourceType();
+      var bound = PreTypeConstraints.Pad($"{adjustableType.UniqueId} :> {AdjustableType.ToStringAsAdjustableType(sourceType)}", 27);
+      var value = PreTypeConstraints.Pad(AdjustableType.ToStringAsBottom(adjustableType), 20);
+      output.WriteLine($"    {bound}  {value}    {TokDescription()}");
+    }
+  }
+}
+
+class FlowFromType : FlowIntoExpr {
+  private readonly Type source;
+
+  public FlowFromType(Type sink, Type source, IToken tok, string description = "")
+    : base(sink, tok, description) {
+    this.source = source;
+  }
+
+  public FlowFromType(Expression sink, Type source, string description = "")
+    : base(sink, sink.tok, description) {
+    this.source = source;
+  }
+
+  protected override Type GetSourceType() {
+    return source;
+  }
+}
+
+class FlowFromTypeArgument : FlowIntoExpr {
+  private readonly Type source;
+  private readonly int argumentIndex;
+
+  public FlowFromTypeArgument(Expression sink, Type source, int argumentIndex)
+    : base(sink, sink.tok) {
+    Contract.Requires(0 <= argumentIndex);
+    this.source = source;
+    this.argumentIndex = argumentIndex;
+  }
+
+  protected override Type GetSourceType() {
+    var sourceType = source.NormalizeExpand();
+    Contract.Assert(argumentIndex < sourceType.TypeArgs.Count);
+    return sourceType.TypeArgs[argumentIndex];
+  }
+}
+
+class FlowFromComputedType : FlowIntoExpr {
+  private readonly System.Func<Type> getType;
+
+  public FlowFromComputedType(Expression sink, System.Func<Type> getType, string description = "")
+    : base(sink, sink.tok, description) {
+    this.getType = getType;
+  }
+
+  protected override Type GetSourceType() {
+    return getType();
+  }
+}
+
+class FlowBetweenExpressions : FlowIntoExpr {
+  private readonly Expression source;
+
+  public FlowBetweenExpressions(Expression sink, Expression source, string description = "")
+    : base(sink, sink.tok, description) {
+    this.source = source;
+  }
+
+  protected override Type GetSourceType() {
+    return source.Type;
+  }
+}
```

Comment:
This is great code, and it seems like it can be made generic enough to be useful in other context. Should we have a directory where we define a generic abstract interpreter that we'll instantiate here and there?

Reply:
Yes, I like that idea. But I propose that we get this one instance of that abstract interpreter working first, and then consider generalizing it to support other abstract interpretation passes.